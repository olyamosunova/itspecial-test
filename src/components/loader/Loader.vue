<template>
    <div class="loader">
        <img :src="spinner" alt="Loading..."/>
    </div>
</template>

<script lang="ts">
    import Vue from 'vue';
    import spinner from '@/assets/spinner.gif';

    export default Vue.extend({
        name: "Loader",
        data() {
            return {
                spinner
            }
        }
    })
</script>

<style scoped>
    .loader {
        flex-grow: 1;
        display: flex;
        flex-direction: column;
        justify-content: center;
        align-items: center;
        padding: 15px;
    }
</style>
