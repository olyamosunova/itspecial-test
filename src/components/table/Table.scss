@import "./src/styles/_variables.scss";
@import "./src/styles/_mixins.scss";

table {
  width: 100%;
  margin: 20px 0;

  border-collapse: collapse;

  @include mediaMax($mobile-max-width) {
    margin: 10px 0;
  }

  thead th {
    vertical-align: middle;
    color: #495057;

    background-color: #e9ecef;
    border-bottom: 1px solid $border-color;

    &.sort {
      cursor: pointer;
    }
  }

  th, td {
    padding: 10px;

    text-align: left;
    white-space: nowrap;
    vertical-align: middle;

    border-top: 1px solid $border-color;
  }

  tbody tr {
    &:hover {
      background-color: #F4F4F4;
    }
  }
}
