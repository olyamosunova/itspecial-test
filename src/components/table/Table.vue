<template>
    <div>
        <div class="overflow-auto mt-5 mb-5">
            <table class="table table-hover">
                <thead class="thead-light">
                <tr>
                    <FieldTitle v-for="field in fields" :key="field.key" :field="field" />
                </tr>
                </thead>
                <tbody>
                <tr
                        v-for="item in data.slice(perPage * (currentPage - 1), perPage * currentPage)"
                        :key="item.id + item.firstName"
                        @click="onClickChooseUser(item.id)"
                >
                    <td>{{item.id}}</td>
                    <td>{{item.firstName}}</td>
                    <td>{{item.lastName}}</td>
                    <td>{{item.phone}}</td>
                    <td>{{item.email}}</td>
<!--                    <td>{{item.adress.city}}, {{item.adress.streetAddress}}, {{item.adress.zip}}, {{item.adress.state}}</td>-->
<!--                    <td>{{item.description}}</td>-->
                </tr>
                <tr>
                    <td v-if="!data.length" colspan="6">No results were found for your search.</td>
                </tr>
                </tbody>
            </table>
        </div>
        <Pagination
                v-if="pages.length > 1"
                :pages="pages"
                @onClickChangePage="onClickChangePage" />
    </div>
</template>

<script>
    import FieldTitle from "../field-title/Field-title";
    import Pagination from "../pagination/Pagination";

    export default {
        name: "Table",
        data() {
            return {
                fields: [
                    {
                        key: 'id',
                        label: 'ID',
                        sortable: true
                    },
                    {
                        key: 'firstName',
                        label: 'First Name',
                        sortable: true
                    },
                    {
                        key: 'lastName',
                        label: 'Last Name',
                        sortable: true
                    },
                    {
                        key: 'phone',
                        label: 'Phone',
                        sortable: false
                    },
                    {
                        key: 'email',
                        label: 'Email',
                        sortable: false
                    },
                    // {
                    //     key: 'adress',
                    //     label: 'Address',
                    //     sortable: false
                    // },
                    // {
                    //     key: 'description',
                    //     label: 'Description',
                    //     sortable: false
                    // }
                ]
            }
        },
        computed: {
            data() {
                return this.$store.getters.data;
            },
            perPage() {
                return this.$store.getters.perPage;
            },
            currentPage() {
                return this.$store.getters.currentPage;
            },
            pagesCount() {
                const length = this.data.length;

                let count = length / this.perPage;

                if (length % this.perPage) {
                    count += 1;
                }

                return count;
            },
            pages() {
                const pages = [];
                for (let i = 1; i <= this.pagesCount; i++) {
                    pages.push(i);
                }

                return pages;
            },
        },
        methods: {
          onClickChangePage({type, page}) {
              switch (type) {
                  case `prev`:
                      this.$store.dispatch('changeCurrentPage', this.currentPage - 1);
                      break;

                  case 'next':
                      this.$store.dispatch('changeCurrentPage', this.currentPage + 1);
                      break;

                  case 'first':
                      this.$store.dispatch('changeCurrentPage', 1);
                      break;

                  case 'last':
                      this.$store.dispatch('changeCurrentPage', Math.trunc(this.pagesCount));
                      break;

                  default:
                      this.$store.dispatch('changeCurrentPage', page);
                      break;
              }
          },
            onClickChooseUser(userId) {
              this.$store.dispatch('getChosenUser', userId);

              setTimeout(() => {
                  window.scrollTo({
                      top: document.body.scrollHeight,
                      behavior: 'smooth'
                  });
              }, 100);
            },
        },
        components: {
            FieldTitle,
            Pagination
        }
    }
</script>

<style lang="scss">
    .table thead th {
        vertical-align: middle;
    }

    th.sort {
        cursor: pointer;
    }

    tbody tr {
        cursor: pointer;
    }
</style>
