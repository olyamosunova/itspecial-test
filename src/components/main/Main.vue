<template>
    <section class="container">
        <FormFilter />

        <Loader v-if="!data" />
        <Table v-else />

        <UserInfo v-if="chosenUser" />
    </section>
</template>

<script>
    import FormFilter from "../form-filter/Form-filter";
    import Table from "../table/Table";
    import Loader from "../loader/Loader";
    import UserInfo from "../user-info/User-info";

    export default {
        name: "Main",
        mounted() {
            this.$store.dispatch('getData');
        },
        computed: {
            data() {
                return this.$store.getters.data;
            },
            chosenUser() {
                return this.$store.getters.chosenUser;
            }
        },
        components: {
            Loader,
            FormFilter,
            Table,
            UserInfo
        },
    }
</script>

<style scoped>

</style>
