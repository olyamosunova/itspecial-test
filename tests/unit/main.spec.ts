import { mount, createLocalVue } from "@vue/test-utils";
import Main from "@/components/main/Main.vue";
import Table from "@/components/table/Table.vue";
import Loader from "@/components/loader/Loader.vue";
import Vuex from "vuex";

const localVue = createLocalVue();
localVue.use(Vuex);

const data = [
    {
        id: 261,
        firstName: "Meribeth",
        lastName: "Haglund",
        email: "VDaly@odio.io",
        phone: "(411)804-5726",
        adress: {
            streetAddress: "9749 Ante Ct",
            city: "El Dorado Hills",
            state: "NV",
            zip: "71328"
        },
        description: "aliquam lacus odio magna lacus sed sed lacus sed convallis ipsum sollicitudin libero magna hendrerit tortor magna porta orci sed aliquam sed vitae magna egestas placerat curabitur amet suspendisse lacus sagittis sollicitudin"
    },
    {
        id: 295,
        firstName: "Marcus",
        lastName: "Espinosa",
        email: "LDover@mattis.com",
        phone: "(497)561-3057",
        adress: {
            streetAddress: "5763 Lorem Ln",
            city: "Walnut Creek",
            state: "TX",
            zip: "92471"
        },
        description: "pharetra sed sit eros augue nec risus porttitor morbi sit sit egestas sit sed dolor pretium vitae magna amet sed rutrum vitae mi amet convallis vitae consectetur massa tincidunt quis sapien vel"
    },
    {
        id: 289,
        firstName: "Ravi",
        lastName: "Kramarsky",
        email: "AGerver@sed.ly",
        phone: "(484)242-1514",
        adress: {
            streetAddress: "8851 Placerat St",
            city: "Oakland",
            state: "UT",
            zip: "80078"
        },
        description: "mattis vel sed ante dui nunc in placerat ipsum vitae nunc morbi sed adipiscing ac vel sollicitudin orci pulvinar sed amet dolor massa vitae turpis sit ac et massa augue tortor amet"
    }
];

describe('unit tests for Main component', () => {
    let actions: any;
    let getters: any;
    let store: any;

    beforeEach(() => {
        actions = {
            getData: jest.fn(),
        };

        getters = {
            data: () => data,
            chosenUser: () => data[0]
        };

        store = new Vuex.Store({
            actions,
            getters
        });
    });

    it('should match the snapshot', () => {
        actions = {
            getData: jest.fn(),
        };

        getters = {
            data: () => data,
            chosenUser: () => data[0]
        };

        store = new Vuex.Store({
            actions,
            getters
        });

        const wrapper = mount(Main, { store, localVue });
        expect(wrapper.html()).toMatchSnapshot();
    });

    it ('should be a Table when data is', () => {
        const wrapper = mount(Main, { store, localVue });
        const table = wrapper.findComponent(Table);
        expect(table.is(Table)).toBe(true)
    });

    it ('should be a Loader when data is empty', () => {
        actions = {
            getData: jest.fn(),
        };

        getters = {
            data: () => null,
            chosenUser: () => null
        };

        store = new Vuex.Store({
            actions,
            getters
        });

        const wrapper = mount(Main, { store, localVue });
        const loader = wrapper.findComponent(Loader);
        expect(loader.is(Loader)).toBe(true);
    });
});
