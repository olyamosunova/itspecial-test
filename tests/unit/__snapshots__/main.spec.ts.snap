// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit tests for Main component should match the snapshot 1`] = `
<section class="container">
  <form class="form">
    <div class="form__block"><input type="text" placeholder="Enter text" aria-label="Enter text" required="required" class="form__control"> <button type="submit" class="button">Find</button></div>
  </form>
  <!---->
  <div>
    <div class="overflow-auto">
      <table class="table">
        <thead>
          <tr>
            <th class="sort">
              ID
              <span class="sort-icon"></span>
            </th>
            <th class="sort">
              First Name
              <span class="sort-icon"></span>
            </th>
            <th class="sort">
              Last Name
              <span class="sort-icon"></span>
            </th>
            <th class="">
              Phone
              <!---->
            </th>
            <th class="">
              Email
              <!---->
            </th>
          </tr>
        </thead>
        <tbody>
          <!---->
        </tbody>
      </table>
    </div>
    <!---->
  </div>
  <div class="card">
    <div class="card__body">
      <p class="card__text">Selected user: <b>Meribeth Haglund</b></p>
      <p class="card__text"> Description:
        <textarea disabled="disabled" class="card__textarea" style="height: 2px;"></textarea>
      </p>
      <p class="card__text">Residence address: <b>9749 Ante Ct</b></p>
      <p class="card__text">City: <b>El Dorado Hills</b></p>
      <p class="card__text">Province / State: <b>NV</b></p>
      <p class="card__text">Zip: <b>71328</b></p>
    </div>
  </div>
  <!---->
</section>
`;
